src/util/user-dirs.js

```diff
@@ -0,0 +1,54 @@
+/* @flow */
+
+const path = require('path');
+const userHome = require('./user-home-dir').default;
+
+const FALLBACK_CONFIG_DIR = path.join(userHome, '.config', 'yarn');
+const FALLBACK_CACHE_DIR = path.join(userHome, '.cache', 'yarn');
+
+export function getDataDir(): string {
+  if (process.platform === 'win32') {
+    const WIN32_APPDATA_DIR = getLocalAppDataDir();
+    return WIN32_APPDATA_DIR == null ? FALLBACK_CONFIG_DIR : path.join(WIN32_APPDATA_DIR, 'Data');
+  } else if (process.env.XDG_DATA_HOME) {
+    return path.join(process.env.XDG_DATA_HOME, 'yarn');
+  } else {
+    // This could arguably be ~/Library/Application Support/Yarn on Macs,
+    // but that feels unintuitive for a cli tool
+
+    // Instead, use our prior fallback. Some day this could be
+    // path.join(userHome, '.local', 'share', 'yarn')
+    // or return path.join(WIN32_APPDATA_DIR, 'Data') on win32
+    return FALLBACK_CONFIG_DIR;
+  }
+}
+
+export function getCacheDir(): string {
+  if (process.platform === 'win32') {
+    // process.env.TEMP also exists, but most apps put caches here
+    return path.join(getLocalAppDataDir() || path.join(userHome, 'AppData', 'Local', 'Yarn'), 'Cache');
+  } else if (process.env.XDG_CACHE_HOME) {
+    return path.join(process.env.XDG_CACHE_HOME, 'yarn');
+  } else if (process.platform === 'darwin') {
+    return path.join(userHome, 'Library', 'Caches', 'Yarn');
+  } else {
+    return FALLBACK_CACHE_DIR;
+  }
+}
+
+export function getConfigDir(): string {
+  if (process.platform === 'win32') {
+    // Use our prior fallback. Some day this could be
+    // return path.join(WIN32_APPDATA_DIR, 'Config')
+    const WIN32_APPDATA_DIR = getLocalAppDataDir();
+    return WIN32_APPDATA_DIR == null ? FALLBACK_CONFIG_DIR : path.join(WIN32_APPDATA_DIR, 'Config');
```

Comment:
If we want to be the "most correct" here, I think the config should go in AppData, while the cache should go in LocalAppData. The difference is that LocalAppData is for anything that should **not** be saved to a roaming profile (ie. synchronised to a server), while regular (roaming) AppData is anything that could/should be synchronised.